streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java

```diff
@@ -16,349 +16,173 @@
                             final String userEndPoint) {
-        this(version, LATEST_SUPPORTED_VERSION, processId, prevTasks, standbyTasks, userEndPoint);
-
-        if (version < 1 || version > LATEST_SUPPORTED_VERSION) {
-            throw new IllegalArgumentException("version must be between 1 and " + LATEST_SUPPORTED_VERSION
-                + "; was: " + version);
+        validateVersions(version, latestSupportedVersion);
+        final SubscriptionInfoData data = new SubscriptionInfoData();
+        data.setVersion(version);
+        if (version >= 3) {
+            data.setLatestSupportedVersion(latestSupportedVersion);
+        }
+        data.setProcessId(processId);
+        data.setPrevTasks(prevTasks.stream().map(t -> {
+            final SubscriptionInfoData.TaskId taskId = new SubscriptionInfoData.TaskId();
+            taskId.setTopicGroupId(t.topicGroupId);
+            taskId.setPartition(t.partition);
+            return taskId;
+        }).collect(Collectors.toList()));
+        data.setStandbyTasks(standbyTasks.stream().map(t -> {
+            final SubscriptionInfoData.TaskId taskId = new SubscriptionInfoData.TaskId();
+            taskId.setTopicGroupId(t.topicGroupId);
+            taskId.setPartition(t.partition);
+            return taskId;
+        }).collect(Collectors.toList()));
+
+        if (version >= 2) {
+            data.setUserEndPoint(userEndPoint == null
+                                     ? new byte[0]
+                                     : userEndPoint.getBytes(StandardCharsets.UTF_8));
         }
+        this.data = data;
     }
 
-    // for testing only; don't apply version checks
-    protected SubscriptionInfo(final int version,
-                               final int latestSupportedVersion,
-                               final UUID processId,
-                               final Set<TaskId> prevTasks,
-                               final Set<TaskId> standbyTasks,
-                               final String userEndPoint) {
-        this.usedVersion = version;
-        this.latestSupportedVersion = latestSupportedVersion;
-        this.processId = processId;
-        this.prevTasks = prevTasks;
-        this.standbyTasks = standbyTasks;
-        this.userEndPoint = userEndPoint;
+    private SubscriptionInfo(final SubscriptionInfoData subscriptionInfoData) {
+        validateVersions(subscriptionInfoData.version(), subscriptionInfoData.latestSupportedVersion());
+        this.data = subscriptionInfoData;
     }
 
     public int version() {
-        return usedVersion;
+        return data.version();
     }
 
     public int latestSupportedVersion() {
-        return latestSupportedVersion;
+        return data.latestSupportedVersion();
     }
 
     public UUID processId() {
-        return processId;
+        return data.processId();
     }
 
     public Set<TaskId> prevTasks() {
-        return prevTasks;
+        return data.prevTasks()
```

Comment:
Would we want to consider building the `Set` the first time this method is invoked then caching for subsequent calls?  Although looking at the code this doesn't seem to be on a hot code path and the task assignments could change, so I'm not sure.  The same goes for the method below.

Reply:
good point. Actually, if my reading is correct, `prevTasks` is actually immutable right here, so we could cache the result.